cajita/unit_test/tstSparseIndexSpace.hpp

```diff
@@ -0,0 +1,681 @@
+                    i_res( i, j, k ) = ti;
+                    j_res( i, j, k ) = tj;
+                    k_res( i, j, k ) = tk;
+                }
+        } );
+    TEST_EXECSPACE().fence();
+    auto i_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), i_res );
+    auto j_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), j_res );
+    auto k_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), k_res );
+    for ( int i = 0; i < size; i++ )
+        for ( int j = 0; j < size; j++ )
+            for ( int k = 0; k < size; k++ )
+            {
+                EXPECT_EQ( i_mirror( i, j, k ), i );
+                EXPECT_EQ( j_mirror( i, j, k ), j );
+                EXPECT_EQ( k_mirror( i, j, k ), k );
+            }
+    bool test[size * size * size];
+
+    Kokkos::View<int[size * size * size], TEST_DEVICE> id_find_res( "id_find" );
+    Kokkos::parallel_for(
+        Kokkos::RangePolicy<TEST_EXECSPACE>( 0, size ), KOKKOS_LAMBDA( int i ) {
+            for ( int j = 0; j < size; j++ )
+                for ( int k = 0; k < size; k++ )
+                {
+                    auto find_id = bis.find( i, j, k );
+                    id_find_res( i * size * size + j * size + k ) = find_id;
+                }
+        } );
+    auto id_find_mirror =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), id_find_res );
+    for ( int i = 0; i < size; i++ )
+        for ( int j = 0; j < size; j++ )
+            for ( int k = 0; k < size; k++ )
+            {
+                test[i * size * size + j * size + k] = false;
+            }
+    for ( int i = 0; i < size; i++ )
+        for ( int j = 0; j < size; j++ )
+            for ( int k = 0; k < size; k++ )
+            {
+                test[id_find_mirror( i * size * size + j * size + k )] = true;
+            }
+    for ( int i = 0; i < size; i++ )
+        for ( int j = 0; j < size; j++ )
+            for ( int k = 0; k < size; k++ )
+            {
+
+                EXPECT_EQ( test[i * size * size + j * size + k], true );
+            }
+}
+
+void testSparseIndexSpaceFullInsert()
+{
+    constexpr int dim_n = 3;
+    constexpr int size_per_dim = 16 * 4;
+
+    std::array<int, dim_n> size( { size_per_dim, size_per_dim, size_per_dim } );
+    int capacity = size_per_dim * size_per_dim;
+    float rehash_factor = 1.5;
+    SparseIndexSpace<TEST_EXECSPACE> sis( size, capacity, rehash_factor );
+
+    auto cbd = sis.cell_bits_per_tile_dim;
+    EXPECT_EQ( cbd, 2 );
+
+    auto cnd = sis.cell_num_per_tile_dim;
+    EXPECT_EQ( cnd, 4 );
+
+    auto cmd = sis.cell_mask_per_tile_dim;
+    EXPECT_EQ( cmd, 3 );
+
+    auto cbt = sis.cell_bits_per_tile;
+    EXPECT_EQ( cbt, 6 );
+
+    auto cnt = sis.cell_num_per_tile;
+    EXPECT_EQ( cnt, 64 );
+
+    Kokkos::View<int***, TEST_DEVICE> qid_res( "query_id", size[0], size[1],
+                                               size[2] );
+
+    Kokkos::parallel_for(
```

Comment:
should check the result of `size()` and `capacity()` after the insert to check for correctness. Can you set the initial capacity/rehash factor such that this insertion test would trigger a rehash? That way the capacity should change before and after the insert procedure.

Reply:
As I mentioned in the previous comments, the rehash factor doesn't really mean the capacity of the hash table. It's for allocation. The capacity of the hash table is initialized as the size of the full grid domain. Looking forward to your suggestions on this setting.